airflow/sensors/base.py

```diff
@@ -251,7 +251,7 @@ def run_duration() -> float:
                 raise AirflowRescheduleException(reschedule_date)
             else:
                 time.sleep(self._get_next_poke_interval(started_at, run_duration, try_number))
-                try_number += 1
+            try_number += 1
```

Comment:
Hi @hussein-awala, this is just put the try_number outside the if logic. For now, if reschedule is true, the `try_number` will always be 1 and the exponential_backoff will be of no use. The exponential_backoff parameter is used in the function: `_get_next_poke_interval`  and it will use `try_number` to generate the backoff interval. If the `try_number`  remains 1, the interval will not change at all, even if the exponential_backoff is set to True. Show some code scripts below.

```
# code in the execute()
if self.reschedule:
    next_poke_interval = self._get_next_poke_interval(started_at, run_duration, try_number)
    reschedule_date = timezone.utcnow() + timedelta(seconds=next_poke_interval)
    if _is_metadatabase_mysql() and reschedule_date > _MYSQL_TIMESTAMP_MAX:
        raise AirflowSensorTimeout(
            f"Cannot reschedule DAG {log_dag_id} to {reschedule_date.isoformat()} "
            f"since it is over MySQL's TIMESTAMP storage limit."
        )
    raise AirflowRescheduleException(reschedule_date)
else:
    time.sleep(self._get_next_poke_interval(started_at, run_duration, try_number))
    try_number += 1
```

```
# code about determine the poke interval
def _get_next_poke_interval(
        self,
        started_at: datetime.datetime | float,
        run_duration: Callable[[], float],
        try_number: int,
    ) -> float:
        """Using the similar logic which is used for exponential backoff retry delay for operators."""
        if not self.exponential_backoff:
            return self.poke_interval

        min_backoff = int(self.poke_interval * (2 ** (try_number - 2)))
```

Reply:
When self.reschedule is True, `raise AirflowRescheduleException(reschedule_date)` will stop the execution, and the `try_number += 1` is not reached even if it's outside the else block.

Even if we add try...finally, updating the try_number cannot solve the problem, because the try_number is resetting when we re-execute (reschedule) the sensor.

IMO, the try_number state should be stored somewhere in order to use it in the next reschedule (ex XCom)

Reply:
Yes. This solution is by far not complete. I will close the PR - we have few others including #30669 from @hussein-awala  and related issue  as well. #18080

 I beleive (correct me if I am wrong @hussein-awala @uranusjr ) but we are converging on the need to add a new entity to keep try_nums ?).